lib/control/control-server-unix.c

```diff
@@ -79,7 +79,13 @@ static void
 control_connection_unix_update_watches(ControlConnection *s)
 {
   ControlConnectionUnix *self = (ControlConnectionUnix *)s;
-  if (s->output_buffer->len > s->pos)
+
+  if (s->waiting_for_output)
+    {
+      iv_fd_set_handler_out(&self->control_io, NULL);
```

Comment:
Can you elaborate? update_watches() is a virtual function that gets called from a number of call-sites.

Where this condition matters is when a command didn't return data.

Anyway, I had some homework to do on this branch, I'll see if  can simplify together with that.

Reply:
Yes I did no explain too well. My intention was to avoid introducing a new bool variable (waiting_for_output) to delay the reply.

I feel uncomfortable with it because seemingly it controls a state, but practically it decides if a command sends replies instantly or delayed, so this flag implicitely carries type information. This argument will not be relevant probably anymore, once every command will be responsible control_connection_send_reply, so my bet this variable would disappear anyway.
Another confusion for me is that I expected some delay involved with this variable (g_cond_wait or anything), but as I see it is only used to "silence" the communication temporarily: `iv_fd_set_handler_out` set to null, so there is no waiting involved at all. (this might be resolved by renaming?) Off topic: why do we need to set the output handler to null, while we waiting?

My original suggestion for change was inspired by your commit message:
```
In case the reply to a command is NULL, we would stop processing the control
socket until a followup call to control_connection_send_reply() is invoked.
```
so the true-falseness of this variable is directly controlled by the reply null-ness, which is equivalent to that the self->output_buffer is empty (modulo a small refactor I mentioned above). Checking output_buffer would eliminate the need of this bool variable, and aligns more with silencing the socket communication.

Let's get to these back later because I suspect the landscape of the patch will change much after the "all commands should call reply" change.

Reply:
Oh I understand now. Thanks!